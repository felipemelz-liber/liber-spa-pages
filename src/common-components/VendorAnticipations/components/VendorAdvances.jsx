import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { CompleteTable, ToastContainer } from 'liber-components';
import {
  TitleContainer,
  ScreenTitle,
  LoadingBlock,
  Loading,
  DisableBootstrap,
  HeaderText,
  HelpLink,
  Container,
  SubTitle,
} from './VendorAdvances.styles';
import { updatePagination, fetchAdvances } from '../actions/VendorAdvancesActions';
import { COLUMNS_ADVANCE_REQUESTED, getColumns } from '../utils';
import PendingAdvances from './PendingAdvances/PendingAdvances';
import Filters from './Filters/Filters';
import usePageState from '../../../hooks/usePageState';
import HideFromAdmin from './HideFromAdmin/HideFromAdmin';

const getHeaderText = (concludedAdvances, pendingAdvances) => {
  if (concludedAdvances.length === 0 && pendingAdvances.length === 0) {
    return (
      <>
        Não encontramos nenhum registro de antecipações realizadas. <br />
        Verifique em <HelpLink href="/fornecedor/duplicatas">Títulos</HelpLink> se é possível fazer
        a sua primeira antecipação.
      </>
    );
  }
  if (pendingAdvances.length === 0) {
    return (
      <>
        Não encontramos nenhuma solicitação em andamento.
        <br /> Confira abaixo o seu histórico de antecipações.
      </>
    );
  }

  return (
    <>
      Encontramos {pendingAdvances.length} solicitação em andamento.
      <br /> Verifique se existem ações pendentes ou confira o histórico de antecipações.
    </>
  );
};

export const VendorAdvances = ({
  concludedAdvances,
  pendingAdvances,
  pagination,
  onUpdatePagination,
  onFetchAdvances,
  loading,
  filters,
  isAdmin,
  onDetailsClick,
}) => {
  const { current, per, pages, count } = pagination;

  const goToPage = selectedPage => {
    onUpdatePagination({ per, page: selectedPage });
  };

  const [pageState, setPageState, onKeyDownPage] = usePageState(current, pages, goToPage);

  React.useEffect(() => {
    onFetchAdvances('pending');
  }, []);

  React.useEffect(() => {
    onFetchAdvances('concluded');
  }, [current, per, filters]);

  const listingStart = () => String((current - 1) * per + 1);
  const listingEnd = () => {
    const end = (current - 1) * per + per;
    if (count > end) return String(end);
    return String(count);
  };

  const handleChangePagination = paginationParams => {
    onUpdatePagination(paginationParams);
  };

  const handlePageInputChange = ({ target: { value } }) => {
    setPageState(value);
  };

  return (
    <>
      <DisableBootstrap />
      <LoadingBlock loading={loading}>
        {loading ? <Loading loop /> : null}
        <Container>
          <HideFromAdmin>
            <TitleContainer>
              <ScreenTitle>Antecipações</ScreenTitle>
            </TitleContainer>
            <HeaderText>{getHeaderText(concludedAdvances, pendingAdvances)}</HeaderText>
            {/* TODO: commented until FAQ exists
            <HelpLink>
              Saiba mais sobre as antecipações
            </HelpLink>
            */}
          </HideFromAdmin>
          <PendingAdvances onDetailsClick={onDetailsClick} isAdmin={isAdmin} />
          {pendingAdvances.length > 0 ? <SubTitle>Histórico de Antecipações</SubTitle> : null}
          <CompleteTable
            items={concludedAdvances}
            columns={getColumns(isAdmin, onDetailsClick)}
            filters={<Filters />}
            paginationProps={{
              currentPage: String(pageState),
              pageTotal: String(pages),
              onChangePage: handlePageInputChange,
              onKeyDownPage,
              onClickNext: () => handleChangePagination({ per, page: current + 1 }),
              onClickPrev: () => handleChangePagination({ per, page: current - 1 }),
              hasListing: true,
              listingProps: {
                onSelectListing: selectedListing =>
                  handleChangePagination({
                    per: parseInt(selectedListing, 10),
                    page: 1,
                  }),
                listing: String(per),
                total: String(count),
                listingLabelSingular: 'adiantamento',
                listingLabelPlural: 'adiantamentos',
                showInterval: {
                  start: listingStart(),
                  end: listingEnd(),
                },
              },
            }}
          />
        </Container>
      </LoadingBlock>

      <ToastContainer />
    </>
  );
};

VendorAdvances.propTypes = {
  pagination: PropTypes.shape({
    count: PropTypes.number,
    current: PropTypes.number,
    next: PropTypes.number,
    pages: PropTypes.number,
    per: PropTypes.number,
    previous: PropTypes.number,
  }).isRequired,
  onUpdatePagination: PropTypes.func,
  concludedAdvances: PropTypes.arrayOf(PropTypes.object),
  pendingAdvances: PropTypes.arrayOf(PropTypes.object),
  filters: PropTypes.shape({}),
  onFetchAdvances: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  onDetailsClick: PropTypes.func,
};

VendorAdvances.defaultProps = {
  onUpdatePagination: () => null,
  onDetailsClick: () => null,
  concludedAdvances: [],
  pendingAdvances: [],
  filters: {},
};

const mapStateToProps = ({
  vendorAdvances: { pagination, concludedAdvances, pendingAdvances, loading, filters, isAdmin },
}) => ({
  pagination,
  concludedAdvances,
  pendingAdvances,
  loading,
  filters,
  isAdmin,
});

const mapDispatchToProps = {
  onUpdatePagination: updatePagination,
  onFetchAdvances: fetchAdvances,
};

export default connect(mapStateToProps, mapDispatchToProps)(VendorAdvances);
