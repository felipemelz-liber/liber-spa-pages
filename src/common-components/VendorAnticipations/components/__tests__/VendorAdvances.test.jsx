import React from 'react';
import sinon from 'sinon';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { VendorAdvances } from '../VendorAdvances';

describe('VendorAdvances component test', () => {
  configure({ adapter: new Adapter() });

  const effectStub = sinon.stub(React, 'useEffect');

  const requiredProps = {
    filters: {},
    pagination: {
      current: 1,
      previous: null,
      next: null,
      per: 10,
      pages: 1,
      count: 0,
    },
    onFetchAdvances: sinon.spy(),
    onSubscribeAdvances: sinon.spy(),
    loading: false,
    onUpdatePagination: () => null,
  };

  const concludedAdvances = [
    {
      number: 0,
    },
  ];

  const pendingAdvances = [
    {
      number: 0,
    },
  ];

  const props = {
    concludedAdvances,
    pendingAdvances,
  };

  afterEach(() => {
    const { onFetchAdvances } = requiredProps;
    onFetchAdvances.resetHistory();
  });

  it(`should render correctly with required props`, () => {
    const wrapper = shallow(<VendorAdvances {...requiredProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it(`should render correctly with all props`, () => {
    const wrapper = shallow(<VendorAdvances {...requiredProps} {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it(`should render correctly with concluded advances and no pending advances`, () => {
    const wrapper = shallow(<VendorAdvances {...requiredProps} {...props} pendingAdvances={[]} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it(`should render correctly with no concluded advances and pending advances`, () => {
    const wrapper = shallow(
      <VendorAdvances {...requiredProps} {...props} concludedAdvances={[]} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should call onFetchAdvances when rendering advances tab', () => {
    const unitedProps = {
      ...requiredProps,
      ...props,
    };
    effectStub.callsArg(0);
    shallow(<VendorAdvances {...unitedProps} />);
    const { onFetchAdvances } = requiredProps;
    expect(onFetchAdvances.calledWith('concluded')).toBeTruthy();
    expect(onFetchAdvances.calledWith('pending')).toBeTruthy();
  });
});
