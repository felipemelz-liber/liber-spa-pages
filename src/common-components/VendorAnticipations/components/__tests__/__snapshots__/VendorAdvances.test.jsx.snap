// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VendorAdvances component test should render correctly with all props 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <styled.div
    loading={false}
  >
    <styled.div>
      <Connect(HideFromAdmin)>
        <styled.div>
          <styled.span>
            Antecipações
          </styled.span>
        </styled.div>
        <styled.span>
          Encontramos 
          1
           solicitação em andamento.
          <br />
           Verifique se existem ações pendentes ou confira o histórico de antecipações.
        </styled.span>
      </Connect(HideFromAdmin)>
      <Connect(PendingAdvances)
        onDetailsClick={[Function]}
      />
      <styled.span>
        Histórico de Antecipações
      </styled.span>
      <CompleteTable
        actions={null}
        allChecked={false}
        borderBottom={false}
        columns={
          Array [
            Object {
              "align": "left",
              "getCellValue": [Function],
              "label": "Status",
              "name": "status",
              "width": 14,
            },
            Object {
              "align": "right",
              "getCellValue": [Function],
              "label": "Valor Antecipado (R$)",
              "name": "netValue",
              "width": 18,
            },
            Object {
              "align": "right",
              "label": "Títulos Solicitados",
              "name": "invoiceCount",
              "width": 15,
            },
            Object {
              "align": "right",
              "getCellValue": [Function],
              "label": "Títulos Antecipados",
              "name": "invoicesAnticipated",
              "width": 15,
            },
            Object {
              "align": "center",
              "label": "Solicitado Em",
              "name": "requestedAt",
              "width": 18,
            },
            Object {
              "align": "center",
              "getCellValue": [Function],
              "label": "Ações",
              "name": "documentUrl",
              "width": 8,
            },
          ]
        }
        emptyState={<Empty />}
        filterBorderTop={true}
        filters={<Memo(Connect(Filters)) />}
        headerBorderTop={true}
        hiddenColumns={Array []}
        isItemSelectable={[Function]}
        isItemSelected={[Function]}
        items={
          Array [
            Object {
              "number": 0,
            },
          ]
        }
        onSelect={[Function]}
        onSelectAll={null}
        paginationProps={
          Object {
            "currentPage": "1",
            "hasListing": true,
            "listingProps": Object {
              "listing": "10",
              "listingLabelPlural": "adiantamentos",
              "listingLabelSingular": "adiantamento",
              "onSelectListing": [Function],
              "showInterval": Object {
                "end": "0",
                "start": "1",
              },
              "total": "0",
            },
            "onChangePage": [Function],
            "onClickNext": [Function],
            "onClickPrev": [Function],
            "onKeyDownPage": [Function],
            "pageTotal": "1",
          }
        }
        selectionControl={null}
        sortingProps={
          Object {
            "onSort": [Function],
            "sortedColumns": false,
            "sorting": Array [],
          }
        }
        tooltipContent={[Function]}
        weightedColumns={false}
      />
    </styled.div>
  </styled.div>
  <_default />
</Fragment>
`;

exports[`VendorAdvances component test should render correctly with concluded advances and no pending advances 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <styled.div
    loading={false}
  >
    <styled.div>
      <Connect(HideFromAdmin)>
        <styled.div>
          <styled.span>
            Antecipações
          </styled.span>
        </styled.div>
        <styled.span>
          Não encontramos nenhuma solicitação em andamento.
          <br />
           Confira abaixo o seu histórico de antecipações.
        </styled.span>
      </Connect(HideFromAdmin)>
      <Connect(PendingAdvances)
        onDetailsClick={[Function]}
      />
      <CompleteTable
        actions={null}
        allChecked={false}
        borderBottom={false}
        columns={
          Array [
            Object {
              "align": "left",
              "getCellValue": [Function],
              "label": "Status",
              "name": "status",
              "width": 14,
            },
            Object {
              "align": "right",
              "getCellValue": [Function],
              "label": "Valor Antecipado (R$)",
              "name": "netValue",
              "width": 18,
            },
            Object {
              "align": "right",
              "label": "Títulos Solicitados",
              "name": "invoiceCount",
              "width": 15,
            },
            Object {
              "align": "right",
              "getCellValue": [Function],
              "label": "Títulos Antecipados",
              "name": "invoicesAnticipated",
              "width": 15,
            },
            Object {
              "align": "center",
              "label": "Solicitado Em",
              "name": "requestedAt",
              "width": 18,
            },
            Object {
              "align": "center",
              "getCellValue": [Function],
              "label": "Ações",
              "name": "documentUrl",
              "width": 8,
            },
          ]
        }
        emptyState={<Empty />}
        filterBorderTop={true}
        filters={<Memo(Connect(Filters)) />}
        headerBorderTop={true}
        hiddenColumns={Array []}
        isItemSelectable={[Function]}
        isItemSelected={[Function]}
        items={
          Array [
            Object {
              "number": 0,
            },
          ]
        }
        onSelect={[Function]}
        onSelectAll={null}
        paginationProps={
          Object {
            "currentPage": "1",
            "hasListing": true,
            "listingProps": Object {
              "listing": "10",
              "listingLabelPlural": "adiantamentos",
              "listingLabelSingular": "adiantamento",
              "onSelectListing": [Function],
              "showInterval": Object {
                "end": "0",
                "start": "1",
              },
              "total": "0",
            },
            "onChangePage": [Function],
            "onClickNext": [Function],
            "onClickPrev": [Function],
            "onKeyDownPage": [Function],
            "pageTotal": "1",
          }
        }
        selectionControl={null}
        sortingProps={
          Object {
            "onSort": [Function],
            "sortedColumns": false,
            "sorting": Array [],
          }
        }
        tooltipContent={[Function]}
        weightedColumns={false}
      />
    </styled.div>
  </styled.div>
  <_default />
</Fragment>
`;

exports[`VendorAdvances component test should render correctly with no concluded advances and pending advances 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <styled.div
    loading={false}
  >
    <styled.div>
      <Connect(HideFromAdmin)>
        <styled.div>
          <styled.span>
            Antecipações
          </styled.span>
        </styled.div>
        <styled.span>
          Encontramos 
          1
           solicitação em andamento.
          <br />
           Verifique se existem ações pendentes ou confira o histórico de antecipações.
        </styled.span>
      </Connect(HideFromAdmin)>
      <Connect(PendingAdvances)
        onDetailsClick={[Function]}
      />
      <styled.span>
        Histórico de Antecipações
      </styled.span>
      <CompleteTable
        actions={null}
        allChecked={false}
        borderBottom={false}
        columns={
          Array [
            Object {
              "align": "left",
              "getCellValue": [Function],
              "label": "Status",
              "name": "status",
              "width": 14,
            },
            Object {
              "align": "right",
              "getCellValue": [Function],
              "label": "Valor Antecipado (R$)",
              "name": "netValue",
              "width": 18,
            },
            Object {
              "align": "right",
              "label": "Títulos Solicitados",
              "name": "invoiceCount",
              "width": 15,
            },
            Object {
              "align": "right",
              "getCellValue": [Function],
              "label": "Títulos Antecipados",
              "name": "invoicesAnticipated",
              "width": 15,
            },
            Object {
              "align": "center",
              "label": "Solicitado Em",
              "name": "requestedAt",
              "width": 18,
            },
            Object {
              "align": "center",
              "getCellValue": [Function],
              "label": "Ações",
              "name": "documentUrl",
              "width": 8,
            },
          ]
        }
        emptyState={<Empty />}
        filterBorderTop={true}
        filters={<Memo(Connect(Filters)) />}
        headerBorderTop={true}
        hiddenColumns={Array []}
        isItemSelectable={[Function]}
        isItemSelected={[Function]}
        items={Array []}
        onSelect={[Function]}
        onSelectAll={null}
        paginationProps={
          Object {
            "currentPage": "1",
            "hasListing": true,
            "listingProps": Object {
              "listing": "10",
              "listingLabelPlural": "adiantamentos",
              "listingLabelSingular": "adiantamento",
              "onSelectListing": [Function],
              "showInterval": Object {
                "end": "0",
                "start": "1",
              },
              "total": "0",
            },
            "onChangePage": [Function],
            "onClickNext": [Function],
            "onClickPrev": [Function],
            "onKeyDownPage": [Function],
            "pageTotal": "1",
          }
        }
        selectionControl={null}
        sortingProps={
          Object {
            "onSort": [Function],
            "sortedColumns": false,
            "sorting": Array [],
          }
        }
        tooltipContent={[Function]}
        weightedColumns={false}
      />
    </styled.div>
  </styled.div>
  <_default />
</Fragment>
`;

exports[`VendorAdvances component test should render correctly with required props 1`] = `
<Fragment>
  <GlobalStyleComponent />
  <styled.div
    loading={false}
  >
    <styled.div>
      <Connect(HideFromAdmin)>
        <styled.div>
          <styled.span>
            Antecipações
          </styled.span>
        </styled.div>
        <styled.span>
          Não encontramos nenhum registro de antecipações realizadas. 
          <br />
          Verifique em 
          <Styled(styled.span)
            href="/fornecedor/duplicatas"
          >
            Títulos
          </Styled(styled.span)>
           se é possível fazer a sua primeira antecipação.
        </styled.span>
      </Connect(HideFromAdmin)>
      <Connect(PendingAdvances)
        onDetailsClick={[Function]}
      />
      <CompleteTable
        actions={null}
        allChecked={false}
        borderBottom={false}
        columns={
          Array [
            Object {
              "align": "left",
              "getCellValue": [Function],
              "label": "Status",
              "name": "status",
              "width": 14,
            },
            Object {
              "align": "right",
              "getCellValue": [Function],
              "label": "Valor Antecipado (R$)",
              "name": "netValue",
              "width": 18,
            },
            Object {
              "align": "right",
              "label": "Títulos Solicitados",
              "name": "invoiceCount",
              "width": 15,
            },
            Object {
              "align": "right",
              "getCellValue": [Function],
              "label": "Títulos Antecipados",
              "name": "invoicesAnticipated",
              "width": 15,
            },
            Object {
              "align": "center",
              "label": "Solicitado Em",
              "name": "requestedAt",
              "width": 18,
            },
            Object {
              "align": "center",
              "getCellValue": [Function],
              "label": "Ações",
              "name": "documentUrl",
              "width": 8,
            },
          ]
        }
        emptyState={<Empty />}
        filterBorderTop={true}
        filters={<Memo(Connect(Filters)) />}
        headerBorderTop={true}
        hiddenColumns={Array []}
        isItemSelectable={[Function]}
        isItemSelected={[Function]}
        items={Array []}
        onSelect={[Function]}
        onSelectAll={null}
        paginationProps={
          Object {
            "currentPage": "1",
            "hasListing": true,
            "listingProps": Object {
              "listing": "10",
              "listingLabelPlural": "adiantamentos",
              "listingLabelSingular": "adiantamento",
              "onSelectListing": [Function],
              "showInterval": Object {
                "end": "0",
                "start": "1",
              },
              "total": "0",
            },
            "onChangePage": [Function],
            "onClickNext": [Function],
            "onClickPrev": [Function],
            "onKeyDownPage": [Function],
            "pageTotal": "1",
          }
        }
        selectionControl={null}
        sortingProps={
          Object {
            "onSort": [Function],
            "sortedColumns": false,
            "sorting": Array [],
          }
        }
        tooltipContent={[Function]}
        weightedColumns={false}
      />
    </styled.div>
  </styled.div>
  <_default />
</Fragment>
`;
